internal/provider/provider.go

```diff
@@ -233,13 +285,133 @@ func (p *pingfederateProvider) Configure(ctx context.Context, req provider.Confi
+			tflog.Info(ctx, "Unable to find password value")
+		} else {
+			hasBasicAuth = true
+		}
+	}
+
+	// Check if the user has provided an access token to the provider
+	var accessToken string
+	var hasAccessTokenAuth bool = false
+	if config.AccessToken.IsUnknown() {
+		// Cannot connect to PingFederate with an unknown value
+		resp.Diagnostics.AddError(
+			"Unable to connect to the PingFederate Server",
+			"Cannot use unknown value as access_token",
+		)
+	} else {
+		if config.AccessToken.IsNull() {
+			accessToken = os.Getenv("PINGFEDERATE_PROVIDER_ACCESS_TOKEN")
+		} else {
+			accessToken = config.AccessToken.ValueString()
+		}
+		if accessToken == "" {
+			tflog.Info(ctx, "Unable to find access_token value")
+		} else {
+			hasAccessTokenAuth = true
+		}
+	}
+
+	// Check if the user has provided an OAuth configuration to the provider
+	var hasOauthConfig bool = false
+	var clientId string
+	var clientSecret string
+	var scopes []string
+	var tokenUrl string
+
+	hasOAuthEnvVars := os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_CLIENT_ID") != "" || os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_CLIENT_SECRET") != "" || os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_TOKEN_URL") != ""
+
+	if internaltypes.IsNonEmptyObj(config.OAuth) || hasOAuthEnvVars {
+		configClientId, ok := config.OAuth.Attributes()["client_id"].(types.String)
+		if ok {
+			clientId = configClientId.ValueString()
+		} else {
+			clientId = os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_CLIENT_ID")
+		}
+
+		configClientSecret, ok := config.OAuth.Attributes()["client_secret"].(types.String)
+		if ok {
+			clientSecret = configClientSecret.ValueString()
+		} else {
+			clientSecret = os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_CLIENT_SECRET")
+		}
+
+		configScopes, ok := config.OAuth.Attributes()["scopes"].(types.List)
+		if ok {
+			configScopes.ElementsAs(ctx, &scopes, false)
+		} else {
+			scopes = []string{os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_SCOPES")}
+		}
+		if len(scopes) == 0 {
+			tflog.Info(ctx, "Unable to find scopes value to be used with OAuth authentication. If this is expected, this information can be ignored.")
+		}
+
+		configTokenUrl, ok := config.OAuth.Attributes()["token_url"].(types.String)
+		if ok {
+			tokenUrl = configTokenUrl.ValueString()
+		} else {
+			tokenUrl = os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_TOKEN_URL")
+		}
+
+		if clientId != "" && clientSecret != "" && tokenUrl != "" {
+			hasOauthConfig = true
+		} else {
+			hasOauthConfig = false
+		}
+	}
+
+	// Validate the configuration
+	// User must provide a username and password, access token, or OAuth configuration to the provider
+	if !hasAccessTokenAuth && !hasBasicAuth && !hasOauthConfig {
+		resp.Diagnostics.AddError(
+			"Unable to find username and password, access_token, or OAuth authentication configuration",
+			"The username and password, access_token, or oauth configuration required values were not supplied. Either set them in the configuration or use the PINGFEDERATE_PROVIDER_* environment variables.",
```

Comment:
I think we are losing some information here. Say they are using oauth but forgot to set tokenUrl. It'd be good if our message says exactly that, but they'd get this more generic message.